scipy/special/meson.build

```diff
@@ -397,36 +399,28 @@ ellint_files = [
 ellint_dep = declare_dependency(sources: ellint_files)
 
 py3.extension_module('_ufuncs_cxx',
-  [ufuncs_cxx_sources, cython_special[2], _cython_tree],
+  [ufuncs_cxx_sources,
+    uf_cython_gen_cpp.process(cython_special[2]),  # _ufuncs_cxx.pyx
+    ],
   cpp_args: [numpy_nodepr_api, use_math_defines],
   include_directories: [inc_np, '../_lib', '../_build_utils/src'],
-  dependencies: [py3_dep, npymath_lib, _ufuncs_pxi_pxd_dep, ellint_dep],
-  override_options: ['cython_language=cpp'],
+  dependencies: [py3_dep, npymath_lib, ellint_dep],
   install: true,
   subdir: 'scipy/special'
 )
 
-cython_special_pxd = declare_dependency(sources: cython_special)
-special_pxd = declare_dependency(sources: _cython_tree)
-
-_ellip_harm_pyx = custom_target('_ellip_harm_pyx',
-  output: '_ellip_harm_2.pyx',
-  input: '_ellip_harm_2.pyx',
-  command: [copier, '@INPUT@', '@OUTDIR@']
-)
-
 py3.extension_module('_ellip_harm_2',
-  [_ellip_harm_pyx, 'sf_error.c'],
+  [uf_cython_gen.process('_ellip_harm_2.pyx'), 'sf_error.c'],
```

Comment:
This fails CI on build warnings like:
```
scipy/special/_ellip_harm_2.cpython-39-x86_64-linux-gnu.so.p/_ellip_harm_2.c:5108:3: error: ‘__pyx_v_t’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
```
`uf_cython_gen` adds `--cplus`, which you don't want here. 

I haven't checked if it's okay to just get rid of `copier` here.

Reply:
I saw the build warnings - but they were puzzling to me.   For that particular case, it's true: https://github.com/scipy/scipy/blob/main/scipy/special/_ellip_harm.pxd#L73 - but - why is this causing an error in this branch, and not previously?

I think you're thinking of `uf_cython_gen_cpp` which does add the `--cplus` flag.

Reply:
Ah, I think it's an inconsistency in build flags. Output from `$ meson introspect build --targets -i` on `main` is:

<details>

```
    {
        "name": "_ellip_harm_2.cpython-39-x86_64-linux-gnu",
        "id": "2b7445d@@_ellip_harm_2.cpython-39-x86_64-linux-gnu@sha",
        "type": "shared module",
        "defined_in": "/home/rgommers/code/scipy/scipy/special/meson.build",
        "filename": [
            "/home/rgommers/code/scipy/build/scipy/special/_ellip_harm_2.cpython-39-x86_64-linux-gnu.so"
        ],
        "build_by_default": true,
        "target_sources": [
            {
                "language": "c",
                "compiler": [
                    "/home/rgommers/anaconda3/envs/scipy-dev/bin/x86_64-conda-linux-gnu-cc"
                ],
                "parameters": [
                    "-I/home/rgommers/code/scipy/build/scipy/special/_ellip_harm_2.cpython-39-x86_64-linux-gnu.so.p",
                    "-I/home/rgommers/code/scipy/build/scipy/special",
                    "-I/home/rgommers/code/scipy/scipy/special",
                    "-I/home/rgommers/anaconda3/envs/scipy-dev/lib/python3.9/site-packages/numpy/core/include",
                    "-I/home/rgommers/code/scipy/build/scipy/_lib",
                    "-I/home/rgommers/code/scipy/scipy/_lib",
                    "-I/home/rgommers/code/scipy/scipy/_build_utils/src",
                    "-I/home/rgommers/code/scipy/build/scipy",
                    "-I/home/rgommers/anaconda3/envs/scipy-dev/include/python3.9",
                    "-I/home/rgommers/anaconda3/envs/scipy-dev/include",
                    "-fdiagnostics-color=always",
                    "-D_FILE_OFFSET_BITS=64",
                    "-w",
                    "-std=c99",
                    "-O2",
                    "-g",
                    "-Wno-unused-but-set-variable",
                    "-Wno-unused-function",
                    "-Wno-conversion",
                    "-Wno-misleading-indentation",
                    "-Wno-incompatible-pointer-types",
                    "-fPIC",
                    "-DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION"
                ],
                "sources": [],
                "generated_sources": [
                    "/home/rgommers/code/scipy/build/scipy/special/_ellip_harm_2.cpython-39-x86_64-linux-gnu.so.p/scipy/special/_ellip_harm_2.pyx.c"
                ]
            },
            {
                "language": "c",
                "compiler": [
                    "/home/rgommers/anaconda3/envs/scipy-dev/bin/x86_64-conda-linux-gnu-cc"
                ],
                "parameters": [
                    "-I/home/rgommers/code/scipy/build/scipy/special/_ellip_harm_2.cpython-39-x86_64-linux-gnu.so.p",
                    "-I/home/rgommers/code/scipy/build/scipy/special",
                    "-I/home/rgommers/code/scipy/scipy/special",
                    "-I/home/rgommers/anaconda3/envs/scipy-dev/lib/python3.9/site-packages/numpy/core/include",
                    "-I/home/rgommers/code/scipy/build/scipy/_lib",
                    "-I/home/rgommers/code/scipy/scipy/_lib",
                    "-I/home/rgommers/code/scipy/scipy/_build_utils/src",
                    "-I/home/rgommers/code/scipy/build/scipy",
                    "-I/home/rgommers/anaconda3/envs/scipy-dev/include/python3.9",
                    "-I/home/rgommers/anaconda3/envs/scipy-dev/include",
                    "-fdiagnostics-color=always",
                    "-D_FILE_OFFSET_BITS=64",
                    "-Wall",
                    "-Winvalid-pch",
                    "-std=c99",
                    "-O2",
                    "-g",
                    "-Wno-unused-but-set-variable",
                    "-Wno-unused-function",
                    "-Wno-conversion",
                    "-Wno-misleading-indentation",
                    "-Wno-incompatible-pointer-types",
                    "-fPIC",
                    "-DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION"
                ],
                "sources": [
                    "/home/rgommers/code/scipy/scipy/special/sf_error.c"
                ],
                "generated_sources": []
            }
        ],
        "extra_files": [],
        "subproject": null,
        "installed": true,
        "install_filename": [
            "/home/rgommers/code/scipy/installdir/lib/python3.9/site-packages/scipy/special/_ellip_harm_2.cpython-39-x86_64-linux-gnu.so"
        ]
    },
```

</details>

And on this PR:

<details>

```
    {
        "name": "_ellip_harm_2.cpython-39-x86_64-linux-gnu",
        "id": "2b7445d@@_ellip_harm_2.cpython-39-x86_64-linux-gnu@sha",
        "type": "shared module",
        "defined_in": "/home/rgommers/code/scipy/scipy/special/meson.build",
        "filename": [
            "/home/rgommers/code/scipy/build/scipy/special/_ellip_harm_2.cpython-39-x86_64-linux-gnu.so"
        ],
        "build_by_default": true,
        "target_sources": [
            {
                "language": "c",
                "compiler": [
                    "/home/rgommers/anaconda3/envs/scipy-dev/bin/x86_64-conda-linux-gnu-cc"
                ],
                "parameters": [
                    "-I/home/rgommers/code/scipy/build/scipy/special/_ellip_harm_2.cpython-39-x86_64-linux-gnu.so.p",
                    "-I/home/rgommers/code/scipy/build/scipy/special",
                    "-I/home/rgommers/code/scipy/scipy/special",
                    "-I/home/rgommers/anaconda3/envs/scipy-dev/lib/python3.9/site-packages/numpy/core/include",
                    "-I/home/rgommers/code/scipy/build/scipy/_lib",
                    "-I/home/rgommers/code/scipy/scipy/_lib",
                    "-I/home/rgommers/code/scipy/scipy/_build_utils/src",
                    "-I/home/rgommers/anaconda3/envs/scipy-dev/include/python3.9",
                    "-I/home/rgommers/anaconda3/envs/scipy-dev/include",
                    "-fdiagnostics-color=always",
                    "-D_FILE_OFFSET_BITS=64",
                    "-Wall",
                    "-Winvalid-pch",
                    "-std=c99",
                    "-O2",
                    "-g",
                    "-Wno-unused-but-set-variable",
                    "-Wno-unused-function",
                    "-Wno-conversion",
                    "-Wno-misleading-indentation",
                    "-Wno-incompatible-pointer-types",
                    "-fPIC",
                    "-DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION"
                ],
                "sources": [
                    "/home/rgommers/code/scipy/scipy/special/sf_error.c"
                ],
                "generated_sources": [
                    "/home/rgommers/code/scipy/build/scipy/special/_ellip_harm_2.cpython-39-x86_64-linux-gnu.so.p/_ellip_harm_2.c"
                ]
            }
        ],
        "extra_files": [],
        "subproject": null,
        "installed": true,
        "install_filename": [
            "/usr/local/lib/python3.9/site-packages/scipy/special/_ellip_harm_2.cpython-39-x86_64-linux-gnu.so"
        ]
    },
```

</details>

Note that `-w` changes to `-Wall` with this PR. So those need suppressing. Probably the approach should be to add `cython_c_args = '-Wno-cpp',  '-Wno-maybe-uninitialized', ...]` and add `cython_c_args` to the `c_args` for generated C code.

Reply:
This is easy to iterate on locally - the build should pass with `python dev.py --build-only --werror`

Reply:
Hmm - thanks for diving into that.  So the`cython` compiler is generating the C build flags for the generated code specifically, which includes, for example, the `-Wno-maybe-uninitialized`?

I guess we're losing functionality if we specify e.g. `-Wno-maybe-uninitialized` for _both_ the generated and the not-generated `.c` file, in that we're losing that check on the not-generated file?

Is there any way to associate C flags with particular files in the sources list?

Reply:
No, they are per target AFAIK. But that's not a big deal, there aren't many places where we mix `.pyx` and `.c` together in interesting ways.